analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Literal.cs

```diff
@@ -0,0 +1,32 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2022 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using StyleCop.Analyzers.Lightup;
+
+namespace SonarAnalyzer.SymbolicExecution.Roslyn.OperationProcessors
+{
+    internal static class Literal
+    {
+        public static ProgramState Process(SymbolicContext context, ILiteralOperationWrapper literal) =>
+            literal.WrappedOperation.ConstantValue is { HasValue: true, Value: null }
+                ? context.State.SetOperationValue(context.Operation, SymbolicValue.Null)
```

Comment:
There will never exist symbolic value on literal, unless we create it by this action. This will safe some memory footprint of the engine.

Reply:
I don't understand how this works. Suppose we have "Null_NotNull"- constraints, a "StringNotEmpty" constraint, and "True_False" constraints. I would have expected three "Process()" invocations for each of the constraints. Maybe like so:
```cs
// Null_NotNullLiteralProcessor.cs
if (constantValue == null) 
  context.State.SetOperationValue(context.Operation, SymbolicValue.Null); 
else
  context.State.SetOperationValue(context.Operation, SymbolicValue.NotNull);
```
```cs
// StringNotEmptyLiteralProcessor.cs
if (constantValue is string s && !string.IsNullOrWhitespace(s))
  context.State.SetOperationValue(context.Operation, SymbolicValue.StringNotEmpty);
```
```cs
// True_FalseLiteralProcessor.cs
if (constantValue is bool b)
  if (b)
    context.State.SetOperationValue(context.Operation, SymbolicValue.True);
  else
    context.State.SetOperationValue(context.Operation, SymbolicValue.False);
```
There are cases where more than one SymbolicValue is present after all three methods are run:
`string x = "hello"; Nullable<bool> x = true;`. With `SetOperationValue` these constraints override each other, don't they?
What am I missing here?

Reply:
None of those cases will ever be combined with `Null` constraint. You cannot have `Null & True`, `Null & Empty`, `Null & NotEmpty` and so on.

If something is `null`, it cannot have any other constraint. So we can optimize the memory and avoid creating duplicate instances. 

We also start with `LiteralOperation`. That is the first ever occurrence of that operation in the execution. So we're at place where we initialize the startup state for it.

Reply:
> If something is null, it cannot have any other constraint.

I'm still concerned here. We know a lot about the left hand side for the null case: it is "definitely assigned", "null or whitespace", and "not true" even though we don't add any of this at the moment.
If you like to optimize, I would still prefer to do it one layer down like e.g. so `AddOrCreateOperationValue(context.Operation, add: ObjectConstraint.Null, create: SymbolicValue.Null);` (inspired by ConcurrentDictionary). This would also be a more universal optimization.
Or each constraint could have its corresponding "StartSymbol":
```cs
    public abstract class SymbolicConstraint
    {
        public abstract SymbolicConstraint Opposite { get; }
        public abstract SymbolicValue StartSymbol { get; } // get's initialized to a singleton  new SymbolicValue().WithConstraint(ObjectConstraint.Null) by ObjectConstraint
    }
```

Reply:
The SE will not work this way.
`definitely assigned` is concern for LVA. SE doesn't care. And SE will know that the symbol has some symbolic value.
`null or whitespace` is not correct. It's `null` and we know it. It cannot be `whitespace`.
`not true` is registered as `false` and we will not build that one.

Same principle was in the [old engine too](https://github.com/SonarSource/sonar-dotnet/blob/58d97fe17231781e37279f3fc6852d9562cbd149/analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Sonar/SymbolicValue.cs#L36) and is proven to work (even though it was a bit overimplemented and underused in that engine)